Tests/TensorFlowTests/EpochsTests.swift

```diff
@@ -0,0 +1,383 @@
+      accessed_.lazy.map(\.accessed)
+    }
+  }
+    
+  func testBaseUse() {
+    let batchSize = 64
+    let dataset = (0..<512).map { (_) -> Tensor<Float> in 
+      Tensor<Float>(randomNormal: [224, 224, 3]) 
+    } 
+    let batches = dataset.inBatches(of: batchSize).lazy.map(\.collated)
+
+    XCTAssertEqual(batches.count, dataset.count / batchSize,
+                   "Incorrect number of batches.")
+    for batch in batches {
+      XCTAssertEqual(batch.shape, TensorShape([64, 224, 224, 3]),
+        "Wrong shape for batch: \(batch.shape), should be [64, 224, 224, 3]")
+    }
+  }
+    
+  func testInBatchesIsLazy() {
+    let batchSize = 64
+    let items = Array(0..<512)
+    let dataset = ReadTracker(items)
+    let batches = dataset.inBatches(of: batchSize)
+      
+    // `inBatches` is lazy so no elements were accessed.
+    XCTAssert(dataset.accessed.allSatisfy(){ !$0 },
+              "No elements should have been accessed yet.")
+    for (i, batch) in batches.enumerated() {
+      // Elements are not accessed until we do something with `batch` so only
+      // the elements up to `i * batchSize` have been accessed yet.
+      XCTAssert(dataset.accessed[..<(i * batchSize)].allSatisfy(){ $0 },
+        "Not all elements prior to \(i * batchSize) have been accessed.")
+      XCTAssert(dataset.accessed[(i * batchSize)...].allSatisfy(){ !$0 },
+        "Some elements after \(i * batchSize) have been accessed.")
+      let _ = Array(batch)
+      let limit = (i + 1) * batchSize
+      // We accessed elements up to `limit` but no further.
+      XCTAssert(dataset.accessed[..<limit].allSatisfy(){ $0 },
+                "Not all elements prior to \(limit) have been accessed.")
+      XCTAssert(dataset.accessed[limit...].allSatisfy(){ !$0 },
+                "Some elements after \(limit) have been accessed.")
+    }
+  }
+
+  func testTrainingEpochsShuffles() {
+    let batchSize = 64
+    let dataset = Array(0..<512)
+    let epochs = TrainingEpochs(samples: dataset, batchSize: batchSize, 
+                                entropy: rng).prefix(10)
+    var lastEpochSampleOrder = Array(0..<512)
+    for batches in epochs {
+      var newEpochSampleOrder: [Int] = []
+      for batch in batches {
+        XCTAssertEqual(batches.count, 8, "Incorrect number of batches.")
+        let samples = Array(batch)
+        XCTAssertEqual(samples.count, batchSize,
+                       "This batch doesn't have batchSize elements.")
+
+        newEpochSampleOrder += samples
+      }
+      XCTAssertNotEqual(
+        lastEpochSampleOrder, newEpochSampleOrder,
+        "Dataset should have been reshuffled.")
+
+      lastEpochSampleOrder = newEpochSampleOrder
+      let uniqueSamples = Set(lastEpochSampleOrder)
+      XCTAssertEqual(
+        uniqueSamples.count, lastEpochSampleOrder.count,
+        "Every epoch sample should be drawn from a different input sample.")
+    }
+  }
+
+  func testTrainingEpochsDropsRemainder() {
```

Comment:
```suggestion
  func testTrainingEpochsShapes() {
```

This seems to test a lot more than just the remainder dropping.